### PR TITLE
Allow entity codecs access to the RequestContext (1.x)

### PR DESCRIPTION
**Pull request by rouzwawi**
Backport of #247 for the `1.x` branch

### CODE REVIEW

File: apollo-entity/src/main/java/com/spotify/apollo/entity/CodecEntityMiddleware.java [hunk 1]
```diff
@@ -19,79 +19,87 @@
...
 
-  CodecEntityMiddleware(EntityCodec codec) {
-    this(codec, codec.defaultContentType());
+  CodecEntityMiddleware(Codec codec) {
+    this.codec = Objects.requireNonNull(codec);
+    this.contentType = null;
   }
 
-  CodecEntityMiddleware(EntityCodec codec, String contentType) {
+  @Deprecated
```
Comment (pettermahlen) [inline apollo-entity/src/main/java/com/spotify/apollo/entity/CodecEntityMiddleware.java]: Maybe add a pointer to a description for the rationale for deprecation, and what should be used instead.

File: apollo-entity/src/main/java/com/spotify/apollo/entity/EntityCodec.java [hunk 2]
```diff
@@ -19,16 +19,15 @@
...
-import com.spotify.apollo.Exploratory;
-
 import java.io.IOException;
-
 import okio.ByteString;
 
 /**
  * Interface for serializing and de-serializing entity types.
+ *
+ * @deprecated Use {@link Codec}
```
Comment (pettermahlen) [inline apollo-entity/src/main/java/com/spotify/apollo/entity/EntityCodec.java]: Add "as that allows for content negotiation" or something similar?

File: apollo-entity/src/test/java/com/spotify/apollo/entity/CodecTest.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+import com.spotify.apollo.Status;
+import com.spotify.apollo.entity.EntityCodecsTest.Entity;
+import com.spotify.apollo.route.Middleware;
+import com.spotify.apollo.route.Route;
+import com.spotify.apollo.test.ServiceHelper;
+import io.norberg.automatter.jackson.AutoMatterModule;
+import okio.ByteString;
+import org.junit.Test;
+
+public class CodecTest {
```
Comment (pettermahlen) [inline apollo-entity/src/test/java/com/spotify/apollo/entity/CodecTest.java]: This is more like `EntityMiddlewareTest`, I think? Codec tests would validate code behaviours, such as whether they perform content negotiation correctly, etc.

Comment (rouzwawi) [reply apollo-entity/src/test/java/com/spotify/apollo/entity/CodecTest.java]: Hmm, there's no actual implementation of content negotiation in the codebase right now. I've done a very simple assertion that the `RequestContext` is being passed into the `Codec` though. I assume that should be enough to allow any code to make decisions on it's own based on that.

Comment (pettermahlen) [reply apollo-entity/src/test/java/com/spotify/apollo/entity/CodecTest.java]: Absolutely, I just meant I don't feel like this class is actually testing codecs, it's testing `EntityMiddleware` ?

Comment (rouzwawi) [reply apollo-entity/src/test/java/com/spotify/apollo/entity/CodecTest.java]: That's true

Comment (rouzwawi) [reply apollo-entity/src/test/java/com/spotify/apollo/entity/CodecTest.java]: Renamed test class

File: apollo-entity/src/main/java/com/spotify/apollo/entity/EntityCodec.java (hunk 2 above)
Comment (rouzwawi) [reply apollo-entity/src/main/java/com/spotify/apollo/entity/EntityCodec.java]: Done

**Pull request merged**